src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +30,109 @@ public static class Renamer
                 newName, optionSet, cancellationToken);
         }
 
+        /// <summary>
+        /// Applies changes to the given document so that the folders, name, and path from <paramref name="newDocumentInfo"/>
+        /// are used. Will also perform symbol changes to the code in the document, attempting to sync type and namespace
+        /// to the new information.
+        /// </summary>
+        /// <returns><see cref="RenameDocumentResult"/> with the modified solution and indication of any symbol changes made</returns>
+        public static async Task<RenameDocumentResult> RenameDocumentAsync(
+            Document document, DocumentInfo newDocumentInfo, OptionSet optionSet, CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            var newName = newDocumentInfo.Name;
+            var shouldRenameType = !string.Equals(document.Name, newName, StringComparison.OrdinalIgnoreCase);
+            var shouldMoveNamespace = !document.Folders.SequenceEqual(newDocumentInfo.Folders);
+
+            var solution = document.Project.Solution
+                .WithDocumentName(document.Id, newDocumentInfo.Name)
+                .WithDocumentFilePath(document.Id, newDocumentInfo.FilePath)
+                .WithDocumentFolders(document.Id, newDocumentInfo.Folders);
+
+            if (!(shouldRenameType || shouldMoveNamespace))
+            {
+                return new RenameDocumentResult(solution, renameSymbolSucceeded: false);
```

Comment:
now i don't like this bool.  `succeeded = false` makes it feel like it failed.  When, in reality this is really just saying "we didn't perform the rename".  I think we might want more fine-grained information about what happened.

Reply:
Consolidating feedback to one entry point for discussion.